### PR TITLE
[Feature] Add CAE

### PR DESCRIPTION
**Pull request by shiyutang**

### CODE REVIEW

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+_base_: '../_base_/ade20k.yml'
+
+batch_size: 4
```
Comment (juncaipeng) [inline configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: 注释说一下total batch size?

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      max_scale_factor: 2.0
+      scale_step_size: 0.0
+    - type: RandomPaddingCrop
+      crop_size: [512, 512]
+    - type: RandomHorizontalFlip
+    - type: RandomDistort
+    - type: Normalize
+      mean: [0.485, 0.456, 0.406]
+      std: [0.229, 0.224, 0.225]
+    - type: Padding
```
Comment (juncaipeng) [inline configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: 这个Padding是不是多余的？前面明确已经crop为512x512了

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    - type: ResizeByShort
+      short_size: 512
+    - type: Normalize
+      mean: [0.485, 0.456, 0.406]
+      std: [0.229, 0.224, 0.225]
+
+loss:
+  types:
+    - type: CrossEntropyLoss
+  types:
```
Comment (juncaipeng) [inline configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: 这里应该写错了  只需要一个types

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+lr_scheduler:
+  type: PolynomialDecay
+  warmup_iters: 1500
+  warmup_start_lr: 0.0
+  learning_rate: 0.0002
+  end_lr: 0.0
+  power: 1.0
+
+test_config:
```
Comment (juncaipeng) [inline configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: test_config 和 dataset 放到一起吧

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml [hunk 5]
```diff
@@ -0,0 +1,86 @@
+_base_: '../_base_/ade20k.yml'
+
```
Comment (juncaipeng) [inline configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: 配置文件名字  xxx_cae_base_xxx

File: paddleseg/cvlibs/config.py [hunk 6]
```diff
@@ -281,6 +281,24 @@ def optimizer(self) -> paddle.optimizer.Optimizer:
...
             return paddle.optimizer.Adam(lr, parameters=params, **args)
+        elif optimizer_type == 'adamwdl':
+            skip_list = self.model.backbone.no_weight_decay()
+
+            decay_dict = {
+                param.name: not (len(param.shape) == 1 or name.endswith(".bias")
+                                 or name in skip_list)
+                for name, param in self.model.named_parameters()
+            }
+            args['n_layers'] = self.model.backbone.get_num_layers()
```
Comment (juncaipeng) [inline paddleseg/cvlibs/config.py]: 这部分逻辑是不是应该放到optimizers.py中，放到这里太复杂了。 
这部分如果对于adamwdl是通用的，就直接放到adamwdl内部。如果只是适用于cae模型，那就在optimizers.py中定义 adamwdl_cae继承adamwdl，放入这部分逻辑。

File: paddleseg/cvlibs/config.py [hunk 7]
```diff
@@ -293,6 +311,8 @@ def optimizer_args(self) -> dict:
         args = self.dic.get('optimizer', {}).copy()
         if args['type'] == 'sgd':
             args.setdefault('momentum', 0.9)
+        elif args['type'] == 'adamwdl':
+            args.pop('momentum', None)
```
Comment (juncaipeng) [inline paddleseg/cvlibs/config.py]: 这个什么意思，如果只是防止用户定义了momentum，建议删掉这个。

File: paddleseg/cvlibs/optimizers.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    if "blocks" in static_name:
+        idx = static_name.find("blocks.")
+        layer = int(static_name[idx:].split(".")[1])
+        ratio = decay_rate**(n_layers - layer)
+    elif "embed" in static_name:
+        ratio = decay_rate**(n_layers + 1)
+    param.optimize_attr["learning_rate"] *= ratio
+
+
+class AdamWDL(AdamW):
```
Comment (juncaipeng) [inline paddleseg/cvlibs/optimizers.py]: 这个是自己实现还是参考实现的，如果和开源实现过于类似，建议加上说明。

File: paddleseg/models/backbones/cae.py [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import math
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+import numpy as np
```
Comment (juncaipeng) [inline paddleseg/models/backbones/cae.py]: import numpy as np 单独一个模块

下面paddle 和paddleseg的import，分为两个模块

File: paddleseg/models/backbones/cae.py [hunk 10]
```diff
@@ -0,0 +1,593 @@
...
+
+        self.qkv = nn.Linear(dim, dim * 3, bias_attr=False)
+
+        if qkv_bias:
+            self.q_bias = self.create_parameter(
+                shape=([dim]), default_initializer=zeros_)
+            self.v_bias = self.create_parameter(
+                shape=([dim]), default_initializer=zeros_)
+            #self.q_bias = nn.Parameter(paddle.zeros(dim)) 
+            #self.v_bias = nn.Parameter(paddle.zeros(dim))
```
Comment (juncaipeng) [inline paddleseg/models/backbones/cae.py]: 无用代码删掉吧，下同

File: paddleseg/models/backbones/cae.py [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+    if token:
+        sinusoid_table = np.concatenate(
+            [sinusoid_table, np.zeros([1, d_hid])], dim=0)
+    return Tensor(sinusoid_table, dtype=float32).unsqueeze(0)
+
+
+@manager.BACKBONES.add_component
+class CAE(nn.Layer):
+    """
+     CAE with support for patch input
```
Comment (juncaipeng) [inline paddleseg/models/backbones/cae.py]: 注释不符合规范

File: paddleseg/models/upernetcae.py [hunk 12]
```diff
@@ -0,0 +1,347 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+#
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 这个文件名改成upernet_cae?

File: paddleseg/models/upernetcae.py [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddleseg.utils import utils
+from paddleseg.cvlibs import manager
+from paddleseg.models import layers
+
+
+class Identity(nn.Layer):
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 这个paddleseg中有，建议import

File: paddleseg/models/upernetcae.py [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+            self.dropout = None
+
+        self.fpn_bottleneck = layers.ConvBNReLU(
+            len(fpn_inplanes) * channels,
+            channels,
+            3,
+            padding=1,
+            bias_attr=False)
+        #self.conv_last = nn.Sequential(
+        #    layers.ConvBNReLU(
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 无用代码

File: paddleseg/models/upernetcae.py [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+
+    def forward(self, input):
+        return input
+
+
+@manager.MODELS.add_component
+class UPerNetCAE(nn.Layer):
+    """ 
+    Unified Perceptual Parsing for Scene Understanding
+    (https://arxiv.org/abs/1807.10221)
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 注释不符合规范

File: paddleseg/models/upernetcae.py [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+    The UPerNet implementation based on PaddlePaddle.
+    Args:
+        num_classes (int): The unique number of target classes.
+        backbone (Paddle.nn.Layer): Backbone network, currently support Resnet50/101.
+        backbone_indices (tuple): Four values in the tuple indicate the indices of output of backbone.
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 多个输入参数没有加注释

File: paddleseg/models/upernetcae.py [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+    """
+
+    def __init__(self,
+                 num_classes,
+                 backbone,
+                 backbone_indices,
+                 channels,
+                 fpn_dim,
+                 inplane_head,
+                 fpn_inplanes,
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 有channels，dim，inplane，建议统一改为channels

File: paddleseg/models/upernetcae.py [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+                 enable_auxiliary_loss=True,
+                 align_corners=True,
+                 dropout_ratio=0.1,
+                 pretrained=None):
+        super(UPerNetCAE, self).__init__()
+        self._init_fpn(embed_dim=768, patch_size=16)
+        self.backbone = backbone
+        self.backbone_indices = backbone_indices
```
Comment (juncaipeng) [inline paddleseg/models/upernetcae.py]: 需要使用backbone_indices来获取backbone输出feature的通道数。

**Review by juncaipeng — COMMENTED** [review]
感觉还挺多不符合规范，没有详细一一看，建议再完善一下代码

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml (hunk 2 above)
Comment (shiyutang) [reply configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: removed

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml (hunk 1 above)
Comment (shiyutang) [reply configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: done

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml (hunk 3 above)
Comment (shiyutang) [reply configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: done

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml (hunk 4 above)
Comment (shiyutang) [reply configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: done

File: configs/cae/upernet_caebase_ade20k_512x512_160k.yml (hunk 5 above)
Comment (shiyutang) [reply configs/cae/upernet_caebase_ade20k_512x512_160k.yml]: done

File: paddleseg/cvlibs/config.py (hunk 7 above)
Comment (shiyutang) [reply paddleseg/cvlibs/config.py]: done

File: paddleseg/cvlibs/config.py (hunk 6 above)
Comment (shiyutang) [reply paddleseg/cvlibs/config.py]: 这部分逻辑不是通用的，但是需要用到模型的属性，不能直接放到oprimizer，否则optimizer需要传入self.model，破坏了其接口的的一致性，因此，我吧adamwdl_cae单独拿出来了，使其配置独立与adamwdl。

File: paddleseg/cvlibs/optimizers.py (hunk 8 above)
Comment (shiyutang) [reply paddleseg/cvlibs/optimizers.py]: 这个是检测的同学实现的，在ppdet中也有

File: paddleseg/models/backbones/cae.py (hunk 9 above)
Comment (shiyutang) [reply paddleseg/models/backbones/cae.py]: done

File: paddleseg/models/backbones/cae.py (hunk 10 above)
Comment (shiyutang) [reply paddleseg/models/backbones/cae.py]: don e

File: paddleseg/models/backbones/cae.py (hunk 11 above)
Comment (shiyutang) [reply paddleseg/models/backbones/cae.py]: done

File: paddleseg/models/upernetcae.py (hunk 12 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: done

File: paddleseg/models/upernetcae.py (hunk 13 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: done

File: paddleseg/models/upernetcae.py (hunk 14 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: done

File: paddleseg/models/upernetcae.py (hunk 15 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: done

File: paddleseg/models/upernetcae.py (hunk 16 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: don

File: paddleseg/models/upernetcae.py (hunk 17 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: done

File: paddleseg/models/upernetcae.py (hunk 18 above)
Comment (shiyutang) [reply paddleseg/models/upernetcae.py]: feature 的通道数这里没有用到。

File: paddleseg/models/upernet_cae.py [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        backbone (Paddle.nn.Layer): Backbone network, currently support Resnet50/101.
+        backbone_indices (tuple): Four values in the tuple indicate the indices of output of backbone.
+        channels(int): Hidden layer channels of upernet head.
+        fpn_channels(int): The fpn_channels of upernet head.
+        head_channels(int): The inplane of upernet head.
+        channels_fpn(int): The channels_fpn of upernet head.
+        enable_auxiliary_loss (bool, optional): A bool value indicates whether adding auxiliary loss. Default: False.
+        align_corners (bool, optional): An argument of F.interpolate. It should be set to False when the feature size is even,
+            e.g. 1024x512, otherwise it is True, e.g. 769x769. Default: False.
+        dropout_ratio(float): The dropout ratio of upernet head.
```
Comment (juncaipeng) [inline paddleseg/models/upernet_cae.py:44]: 加 optional

File: paddleseg/cvlibs/config.py (hunk 6 above)
Comment (juncaipeng) [reply paddleseg/cvlibs/config.py]: 不能再在这里增加逻辑判断了，要不然config class变得太乱。还是应该在optimizers.py中定义 adamwdl_cae 继承adamwdl，放入这部分逻辑，免去了很多config中的代码。 
optimizer需要传入self.model也没问题，反正这个adamwdl_cae是专用的。

File: paddleseg/models/upernet_cae.py [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+        enable_auxiliary_loss (bool, optional): A bool value indicates whether adding auxiliary loss. Default: False.
+    """
+
+    def __init__(self,
+                 inplane,
+                 num_class,
+                 channels_fpn,
+                 channels,
+                 dropout_ratio=0.1,
+                 fpn_channels=512,
```
Comment (juncaipeng) [inline paddleseg/models/upernet_cae.py:247]: channels_fpn 和 fpn_channels，两个参数有点迷哦，注释也需要补充

**Comment by shiyutang** [comment]
正常训练，损失下降：
image

**Review by juncaipeng — APPROVED** [review]
